src/openloco/windows/LandscapeGeneration.cpp

```diff
@@ -0,0 +1,1371 @@
+                window->x + 10,
+                window->y + window->widgets[widx::start_year].top,
+                colour::black,
+                string_ids::start_year,
+                nullptr);
+        }
+
+        // 0x0043DB76
+        static void prepare_draw(window* window)
+        {
+            common::prepare_draw(window);
+
+            commonFormatArgs[0] = *scenarioStartYear;
+
+            if ((scenarioFlags & scenario::flags::landscape_generation_done) == 0)
+            {
+                window->activated_widgets |= (1 << widx::generate_when_game_starts);
+                window->disabled_widgets |= (1 << widx::generate_now);
+            }
+            else
+            {
+                window->activated_widgets &= ~(1 << widx::generate_when_game_starts);
+                window->disabled_widgets &= ~(1 << widx::generate_now);
+            }
+        }
+
+        static void confirmResetLandscape(int32_t eax)
+        {
+            bool madeAnyChanges = addr<0x009C871C, int32_t>();
+            if (madeAnyChanges)
+            {
+                LandscapeGenerationConfirm::open(eax);
+            }
+            else
+            {
+                WindowManager::close(WindowType::landscapeGenerationConfirm, 0);
+
+                // 0x0043EDAD
+                *scenarioFlags &= ~(scenario::flags::landscape_generation_done);
+                WindowManager::invalidate(WindowType::landscapeGeneration, 0);
+                call(0x0043C88C);
+                addr<0x009C871C, uint8_t>() = 0;
+                addr<0x00F25374, uint8_t>() = 0;
+                gfx::invalidate_screen();
+            }
+        }
+
+        // 0x0043DC83
+        static void on_mouse_down(window* window, widget_index widgetIndex)
+        {
+            switch (widgetIndex)
+            {
+                case widx::start_year_up:
+                    if (*scenarioStartYear + 1 <= scenario::max_year)
+                        *scenarioStartYear += 1;
+                    window->invalidate();
+                    break;
+
+                case widx::start_year_down:
+                    if (*scenarioStartYear - 1 >= scenario::min_year)
+                        *scenarioStartYear -= 1;
+                    window->invalidate();
+                    break;
+            }
+        }
+
+        // 0x0043DC58
+        static void on_mouse_up(window* window, widget_index widgetIndex)
+        {
+            switch (widgetIndex)
+            {
+                case common::widx::close_button:
+                    WindowManager::close(window);
+                    break;
+
+                case common::widx::tab_options:
+                case common::widx::tab_land:
+                case common::widx::tab_forests:
+                case common::widx::tab_towns:
+                case common::widx::tab_industries:
+                    common::switchTab(window, widgetIndex);
+                    break;
+
+                case widx::start_year_up:
```

Comment:
Is not in the original mouse up handler

Reply:
Rebase error; I moved them to `on_mouse_down` once I realised I was looking at both events at once in IDA.